module/zfs/zfs_vnops.c

```diff
@@ -979,13 +981,497 @@ zfs_get_done(zgd_t *zgd, int error)
+	if (error != 0) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (error);
+	}
+
+	if (!spa_feature_is_enabled(dmu_objset_spa(dstzfsvfs->z_os),
+	    SPA_FEATURE_BLOCK_CLONING)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EOPNOTSUPP));
+	}
+
+	/*
+	 * We don't copy source file's flags that's why we don't allow to clone
+	 * files that are in quarantine.
+	 */
+	if (srczp->z_pflags & ZFS_AV_QUARANTINED) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EACCES));
+	}
+
+	/*
+	 * Both source and destination have to belong to the same storage pool.
+	 */
+	if (dmu_objset_spa(srczfsvfs->z_os) !=
+	    dmu_objset_spa(dstzfsvfs->z_os)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EXDEV));
+	}
+
+	if (srcoffset >= srczp->z_size) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (0);
+	}
+	if (length == 0) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (0);
+	}
+	if (length == UINT64_MAX)
+		length = srczp->z_size - srcoffset;
+
+	/*
+	 * Callers might not be able to detect properly that we are read-only,
+	 * so check it explicitly here.
+	 */
+	if (zfs_is_readonly(dstzfsvfs)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EROFS));
+	}
+
+	/*
+	 * If immutable or not appending then return EPERM.
+	 * Intentionally allow ZFS_READONLY through here.
+	 * See zfs_zaccess_common()
+	 */
+	if ((dstzp->z_pflags & ZFS_IMMUTABLE)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EPERM));
+	}
+
+	/*
+	 * XXX: There should be no need to sync the source file, as we return
+	 *      an error if somebody is trying to clone a block that was
+	 *      modified in the same transaction group.
+	 */
+#ifdef FRSYNC
+	/*
+	 * If we're in FRSYNC mode, sync out this znode before reading it.
+	 * Only do this for non-snapshots.
+	 *
+	 * Some platforms do not support FRSYNC and instead map it
+	 * to O_SYNC, which results in unnecessary calls to zil_commit. We
+	 * only honor FRSYNC requests on platforms which support it.
+	 */
+	frsync = !!(srcioflag & FRSYNC);
+#endif
+	if (srczfsvfs->z_log &&
+	    (frsync || srczfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS)) {
+		zil_commit(srczfsvfs->z_log, srczp->z_id);
+	}
+
+	/*
+	 * No overlapping if we are cloning within the same file.
+	 */
+	if (srczp == dstzp) {
+		if ((srcoffset >= dstoffset &&
+		    srcoffset < dstoffset + length) ||
+		    (dstoffset >= srcoffset &&
+		    dstoffset < srcoffset + length)) {
+			zfs_exit_two(srczfsvfs, dstzfsvfs);
+			return (SET_ERROR(EINVAL));
+		}
+	}
+
+	/*
+	 * Maintain predictable lock order.
+	 */
+	if (srczp < dstzp) {
+		srclr = zfs_rangelock_enter(&srczp->z_rangelock, srcoffset,
+		    length, RL_READER);
+		dstlr = zfs_rangelock_enter(&dstzp->z_rangelock, dstoffset,
+		    length, RL_WRITER);
+	} else {
+		dstlr = zfs_rangelock_enter(&dstzp->z_rangelock, dstoffset,
+		    length, RL_WRITER);
+		srclr = zfs_rangelock_enter(&srczp->z_rangelock, srcoffset,
+		    length, RL_READER);
+	}
```

Comment:
Should here also be a case of srczp == dstzp, comparing srcoffset to dstoffset to avoid deadlock if somebody try to do it simultaneously in opposite directions.

Reply:
Good catch!